resip/stack/ConnectionBase.cxx

```diff
@@ -1049,6 +1059,33 @@ ConnectionBase::getWriteBufferForExtraBytes(int extraBytes)
       return mBuffer;
    }
 }
+*/
+
+
+
```

Comment:
Take a look at getCurrentWriteBuffer.  This is what is passed to TlsConnection::read.  Since the buf pointer that TlsConnection::read has can actually be in the middle of the mBuffer, we can't assume that readBytes is the correct index to where the current data ends with respect to the mBuffer pointer.  ie: TlsConnection::read buf pointer != mBuffer.

I'm guessing this is working for you, because mBufferPos is always 0 when getWriteBufferForExtraBytes is called?

Factoring in mBufferPos into your method should address this concern (if I'm understanding all this correctly).  Assuming you agree with the statements above, could be please try out the following version instead and see if it still works for you?

char*
ConnectionBase::getWriteBufferForExtraBytes(int readBytes, int extraBytes)
{
   if (readBytes > 0 && extraBytes > 0)
   {
      if ((mBufferPos + readBytes + extraBytes) > mBufferSize)
      {
         mBufferSize = mBufferPos + readBytes + extraBytes;
         char* buffer = MsgHeaderScanner::allocateBuffer((int)mBufferSize);
         memcpy(buffer, mBuffer, mBufferPos + readBytes);
         delete[] mBuffer;
         mBuffer = buffer;
      }
      return &mBuffer[mBufferPos + readBytes];
   }
   else
   {
      resip_assert(0);
      return mBuffer;
   }
}

Once I know I understand the problem properly, I might recommend a difference approach to this fix, since with these changes the spirit of the getWriteBufferForExtraBytes has really changed and it's not entirely clear what the readBytes parameter is all about without seeing the context of use in the subclassed TlsConnection class.

Reply:
>> I'm guessing this is working for you, because mBufferPos is always 0 when >> >getWriteBufferForExtraBytes is called?

I think it would be more precise to say that the ConnectionState is always NewMessage every time Connection::read()  is called (and mBufferPos gets reset to 0) and that's why it works for me.

I understand your concern about mBufferPos and I had the same concern. But note that the old getWriteBufferForExtraBytes() doesn't seem to care about what mBufferPos's value is, so I left it out of the new calculation as well.

But you may be right about this. My testing has been with  WSS and there may be some kind of restriction in the spec about a single write containing the entire message and that constraint might not exist for SIP-over-regular non-WSS TLS.